modules: remove file extension and directory resolving for esm

MylesBorins opened the pull request:
this build on top of the limitations of the package name map proposal.
It removes file extensions and directory resolution in the resolver.

This is only currently limited for local development. When resolving
dependencies file extension and directory resolution will still work.

Refs: https://github.com/domenic/package-name-maps

Alternative to https://github.com/nodejs/node/pull/21572

MylesBorins commented:
ping @zenparsing I borrowed `ResolvesToFile` from your branch. LMK how you would like to be attributed

https://github.com/nodejs/node/pull/21729/commits/95aad4f0ce4575a6f35dfd28e9c83edeb7851d8b#diff-4a5950be44f56dfca2e7fa3c4e1fc0e4R551

MylesBorins commented on src/module_wrap.cc:
```diff
@@ -632,10 +638,27 @@ Maybe<URL> ResolveDirectory(Environment* env,
 }  // anonymous namespace
 
-Maybe<URL> Resolve(Environment* env,
+Maybe<URL> ResolveRecurse(Environment* env,
```
This was done to make a separate code path for resolving dependencies than local modules. Not 100% happy with the current implementation... but there is definitely a difference in the algorithm now (still need to resolve main for modules).

Open to suggestions on better ways to do this

guybedford replied:
If this is renamed to `ResolveRelativeWithExtensions`, and used only by main lookups, then the `ShouldBeTreatedAsRelativeOrAbsolutePath` check can be skipped.

MylesBorins commented:
@bmeck totally reasonable. I look forward to discussing it. To be explicit I consider this a WIP, and have added the appropriate labels.

bmeck commented:
@MylesBorins to be clear I am fairly -1 to package-name-map parity currently and will probably be needing more than just technical work to not think this is a bad idea.

MylesBorins commented:
@bmeck makes sense to discuss in the WG.

guybedford commented on src/module_wrap.cc:
```diff
@@ -586,9 +592,9 @@ Maybe<URL> ResolveMain(Environment* env, const URL& search) {
   }
   if (!ShouldBeTreatedAsRelativeOrAbsolutePath(pjson.main)) {
-    return Resolve(env, "./" + pjson.main, search, IgnoreMain);
+    return ResolveRecurse(env, "./" + pjson.main, search, IgnoreMain);
```
This can probably be simplified to a `ResolveRelativeWithExtensions`, that doesn't need to do the absolute checks and assumes its input is a relative URL.

guybedford commented on src/module_wrap.cc:
```diff
@@ -599,7 +605,7 @@ Maybe<URL> ResolveModule(Environment* env,
     dir = parent;
     Maybe<URL> check =
-        Resolve(env, "./node_modules/" + specifier, dir, CheckMain);
+        ResolveRecurse(env, "./node_modules/" + specifier, dir, CheckMain);
```
For simplicity, this can stay as just `Resolve`.

guybedford commented on src/module_wrap.cc:
```diff
@@ -650,13 +673,9 @@ Maybe<URL> Resolve(Environment* env,
-    }
-    return ResolveDirectory(env, resolved, check_pjson_main);
+    if (ResolvesToFile(resolved))
+      return Just(resolved);
```
Beautiful :)

MylesBorins replied:
That's all @zenparsing

zenparsing replied:
BTW, this C++ code is a pleasure to work with 👍

guybedford replied:
All the implementation work of @bmeck, with many various refactoring contribs.

MylesBorins commented:
@guybedford wanna push a patch to address your comments? I'll be afk the rest of the day and likely not have a tonight time to work on this this week

guybedford commented:
@MylesBorins also playing holiday catch-up here, but will add it to the list.

guybedford commented:
I've added the changes here, although the latest commit on this branch with some typo corrections isn't being displayed by GitHub for me for some reason... https://github.com/nodejs/node/commit/be9afef5617f9990886d161b5a39c2bc6c9a93e1.

MylesBorins commented:
@nodejs/modules @bmeck is this perhaps something we can move forward with if we get all the tests passing?

mcollina commented:
I’m +1 in getting this landed.

MylesBorins commented:
I've rebased against master and fixed a problem with compilation

@guybedford could you ptal?

CI: https://ci.nodejs.org/job/node-test-pull-request/16310/

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -728,7 +729,11 @@ if (experimentalModules) {
+  const ext = path.extname(base);
+  const isESM = ext === '.mjs';
+
+  if (experimentalModules && (isESM || hasLoader)) {
```
I would prefer if we could move this filtering logic into the resolver itself still, to ensure a single code path for resolution.

guybedford commented on src/module_wrap.cc:
```diff
@@ -605,8 +605,8 @@ Maybe<URL> ResolveMain(Environment* env, const URL& search) {
-  if (resolved.back() != '/') {
-    resolved = URL(resolved + "/", search);
+  if (pjson.main.back() != '/') {
+    resolved = URL(pjson.main + "/", search);
```
This may need a little more work, as I think we're supposed to take into account both this check and the leading `./` check.

guybedford commented on src/module_wrap.cc:
```diff
@@ -623,23 +639,11 @@ Maybe<URL> ResolveModule(Environment* env,
-                            const URL& search,
-                            PackageMainCheck check_pjson_main) {
-  if (check_pjson_main) {
-    Maybe<URL> main = ResolveMain(env, search);
```
It seems this is never called now.

This is likely my fault - rushed the refactoring a bit last time. Will take another look.

bmeck commented:
I would not like to see this land for now. Per my existing requested changes around my thoughts around not wanting to rely on such an alpha spec for package name maps that I also have raised concerns with; as well as comments like https://github.com/nodejs/modules/issues/141#issuecomment-402313612 about ecosystem breakage, I don't think this change is warranted.

If this were not attempting to use such a premature spec as reasoning and that spec itself did not seem problematic to me, discussion of breakage would be the main discourse that I would entail still. Reasoning about needing to match the web when there are repeated discussions about the web not needing to match node leads to a master-slave like scenario that I have no desire to encourage.

demurgos commented:
Despite hoping to have optional extensions in the final product (a must have for a smooth transition IMO), restricting to mandatory extensions seems safe if the goal is to have a "common core".

Still, I thought that the current policy regarding changes to Node's module system was to only accept bug fixes (semver patch changes). Is it an exception?

guybedford commented:
@bmeck the question here isn't "should we disable automatic extensions", it is "might we want to disable automatic extensions". If you don't agree with the former that is fine and things can quite happily still go that way, but the window of opportunity is closing to even have the ability to consider supportting this unless we move on this PR soon.

bmeck commented:
It is my understanding that we are in a moratorium for such changes still, yes. However, lots of people seem to make PRs for discussion and that can make it hard to understand what paths are being moved forward with as a group vs as a proposal by individuals. 

> restricting to mandatory extensions seems safe if the goal is to restrict Node's current ESM support to a common core shared by all proposals.

I actually find this to be a breaking change that *should not* be considered part of a common core. For reasons like you lay out in https://github.com/nodejs/modules/issues/139 ; it has interoperability problems not just with existing code, but also with migrations.

bmeck commented:
> but the window of opportunity is closing to even have the ability to consider supportting this unless we move on this PR soon.

We cannot saddle both sides of "node's implementation is experimental and subject to breaking changes" and "node's implementation is experimental and enforces patterns". My fear here is we are making a big change to the implementation with great impact on a variety of topics without discussion of those.

I don't even see mention on migration patterns for multi-mode packages here? Where is the discussion about the impacts of landing this. Similar to https://github.com/nodejs/node/pull/21317 I don't think this PR is looking at overall impact but instead is scoping its focus to finely to a specific change that makes us lose sight of what this would do to an overall implementation.

I also see an incomplete and what I see as problematic package-name-map spec as a large part of the reasoning here, but as pointed out in the past you can achieve this kind of ["searching" in package name maps](https://github.com/nodejs/modules/issues/51#issuecomment-406058217) so that invalidates this line of reasoning entirely. I think this needs to be more well discussed about impact than just saying it is for some level of compatibility with a potential future web spec that contradicts this PR.

I do consider this change controversial and do not want to see it being discussed without recognizing the impact that this has. Right now it is being phrased as:

 * a simple change but actually has **major** affects on migration.
 * a web compatibility change but actually is contradicted by the early stage spec it is saying it needs the change for
 * a reasonable reservation of future design space but isn't recognizing the ecosystem breakage and the feedback from a minimal implementation PR that tried this

demurgos commented:
@bmeck 
>>  restricting to mandatory extensions seems safe if the goal is to restrict Node's current ESM support to a common core shared by all proposals.
>
> I actually find this to be a breaking change that should not be considered part of a common core. For reasons like you lay out in nodejs/modules#139 ; it has interoperability problems not just with existing code, but also with migrations.

By "common core shared by all proposals", I mean that enforcing mandatory extensions now means that all of the current can be implemented. Optional extensions can be restored later, but having optional extensions rules out any proposal with mandatory.

As I stated above and in my interop issue, I think that mandatory extensions would cause a migration failure and I am strongly opposed to it in the final product. It's just note that until there is consensus, mandatory extensions are the less-commiting choice (you can't remove optional extensions once it's out).

demurgos commented:
@bmeck 
>>  restricting to mandatory extensions seems safe if the goal is to restrict Node's current ESM support to a common core shared by all proposals.
>
> I actually find this to be a breaking change that should not be considered part of a common core. For reasons like you lay out in nodejs/modules#139 ; it has interoperability problems not just with existing code, but also with migrations.

By "common core shared by all proposals", I specifically mean that it does not rule out any proposal. Optional extensions can be restored later, but having optional extensions prevent any proposal where they are mandatory.

As I stated above and in my interop issue, I think that mandatory extensions would cause a migration failure and I am strongly opposed to them in the final product. It's just that until there is consensus, mandatory extensions are the less-commiting choice (you can't remove optional extensions once they are out). I consider landing this PR would be step backward, but I understand why it's pushed by some people.

I am mostly worried because I did not expect the current implementation to change until the discussions were resolved.

guybedford commented:
> We cannot saddle both sides of "node's implementation is experimental and subject to breaking changes" and "node's implementation is experimental and enforces patterns".

We kind of have to accept both to some degree. Anything that works in the experimental implementation will be a painful break at some level if it does not work in the non-experimental implementation. Eg for this reason it is not possible to remove ".mjs" at this point - it is already cemented into the ecosystem (not that we would of course).

> My fear here is we are making a big change to the implementation with great impact on a variety of topics without discussion of those.

Agreed, let's discuss.

> I don't even see mention on migration patterns for multi-mode packages here? Where is the discussion about the impacts of landing this.

Multi-mode support would still work for consumer-side `import 'pkg/x'` and `import 'pkg'`, but not for `import '/path/to/pkg/x'`.

This change of no longer supporting "importing a folder being equivalent to importing a package" is a very important concern definitely, and the loss of a use case in Node.

We could consider a change to this PR to retain the "treat folders as packages" behaviour, while still removing the automat extensions behaviours.

@MylesBorins let me know if you could get behind such a change as well.

@bmeck do please share any other concerns on this.

guybedford commented:
> this change breaks the world for people like me who are working on developing the loader.

@devsnek can you perhaps share some details on how this affects things for you?

devsnek commented:
@guybedford it breaks things because node is barely usable without node-style resolution. even if this is undone at some point in the future I see no reason to kneecap ourselves.

guybedford commented:
> it breaks things because node is barely usable without node-style resolution. even if this is undone at some point in the future I see no reason to kneecap ourselves.

@devsnek could you give an example of what you would expect that wouldn't work here that you deem as this kind of compatibility loss based on my amended proposal in https://github.com/nodejs/node/pull/21729#issuecomment-412074535?

bmeck commented:
> This change of no longer supporting "importing a folder being equivalent to importing a package" is a very important concern definitely, and the loss of a use case in Node.

I'd say it is more impactful than a mere single use case. The divergence in behaviors means things like `resolve('pkg/foo')` is unable to be used as a cached value if the use of it would eventually rely on searching:

```mjs
const foo = resolve('pkg/foo');
import(new URL('./bar', foo));
```

would fail and need to be rewritten as:

```mjs
 // this should be rewritten in a safe way to compose specirier paths
// I'm unclear on how to do this easily right now
const foo = 'pkg/foo';
import(`${foo}/../bar`);
```

but the above actually has a bug in it compared to the original. It imports `pkg/bar` which may not be the same thing that `pkg/foo` pointed to when resolved. For example, `pkg/foo` could have a `package.json` pointing to `pkg/foo/lib/index.js`) leading the original to end up with ~ `pkg/foo/lib/bar` and the rewritten to be `pkg/bar`.

guybedford commented:
@bmeck my amendment in https://github.com/nodejs/node/pull/21729#issuecomment-412074535 would deal with that case as discussed in that comment, so that `import(new URL('./bar', foo));` would still resolve correctly.

bmeck commented:
@guybedford `import(new URL('./bar', foo));` resolves to an absolute URL which seems exactly what you call out as not working:

> not for import '/path/to/pkg/x'.

devsnek commented:
@guybedford when I resolve `./x` I expect to not have to know how x is implemented. if x is 500 files in a directory or a single .json file or whatever doesn't matter. I'm not really sure why this isn't considered like the bare minimum here.

guybedford commented:
@bmeck I'm referring to:

> We could consider a change to this PR to retain the "treat folders as packages" behaviour, while still removing the automatic extensions behaviours.

To try and clarify the approach, we could do a package-boundary check, and then apply extensions only if the parent package boundary is a different package to the resolved package boundary.

GeoffreyBooth commented:
We agreed at the last meeting to work on creating a minimal implementation that could be a core that various implementations could build on. It seems to me that what's in core Node should be that minimal implementation, even if it's not fully usable and doesn't satisfy most use cases. This PR seems to me to be along those lines. I think it's time for the "full" experimental-modules implementation to live in a branch, and what's in core is the minimal base we agreed to build.

GeoffreyBooth commented:
The moratorium shouldn’t extend to PRs. If it does we’ll never make any progress.

devsnek commented:
i'm used to prs meaning someone wants to land something. if that wasn't the intention here then i suppose its fine... a label or title change might not go amiss to make sure people understand it.

GeoffreyBooth commented:
> it doesn't make sense to land them separately without having that story in place.

Perhaps we could make these PRs against some other branch? To keep the PRs minimal but assemble the story elsewhere until it's cohesive and could then be merged into core.

bmeck commented:
@GeoffreyBooth that seems ok. Would put them all to land on that fork/branch and have the overarching discussions before coming back here I would assume. Getting to consensus would still take time as long as that is ok.

benjamingr commented:
> Looking forward to discussing in a meeting; i find extension and directory resolution a critical feature to have by default

@ljharb would you mind elaborating on why?

ljharb commented:
The entire npm ecosystem, including best practices around specifiers, relies on full resolution. Package name maps are hopefully being added to browsers to make them more like node - it would be a massive step backwards for node to lose this feature.

MylesBorins commented:
Moved to https://github.com/nodejs/ecmascript-modules/pull/2
